src/App.tsx

```diff
@@ -1,223 +1,243 @@
       newTodoField.current.focus();
     }
+
+    if (user) {
+      getTodos(user.id)
+        .then(setTodos)
+        .catch(() => {
+          setErrorMessage('Unable to download todos');
+        });
+    }
+  }, []);
+
+  const uncompletedTodosLength = useMemo(() => (
+    todos.filter(todo => !todo.completed).length
+  ), [todos]);
+
+  const completedTodosLength = useMemo(() => (
+    todos.filter(todo => todo.completed).length
+  ), [todos]);
+
+  const visibleTodos = useMemo(() => {
+    return filterTodos(todos, complitedFilter);
+  }, [complitedFilter, todos]);
+
+  const deleteTodo = useCallback(async (todoId: number) => {
+    try {
+      const deleteResponse = await deleteTodoById(todoId);
+
+      setTodos(prevTodos => (
+        prevTodos.filter(todo => todo.id !== todoId)
+      ));
+
+      return deleteResponse;
+    } catch (deleteError) {
+      return setErrorMessage('Unable to delete a todo');
+    }
   }, []);
 
+  const clearCompletedTodos = useCallback(() => {
+    todos.forEach(todo => {
+      if (todo.completed) {
+        deleteTodo(todo.id);
+      }
+    });
+  }, [todos]);
+
+  const addTodo = useCallback(
+    async (event: React.FormEvent<HTMLFormElement>, title: string) => {
+      event.preventDefault();
+
+      try {
+        if (!title.trim()) {
+          setErrorMessage('Title can\'t be empty');
+
+          return;
+        }
+
+        setIsNewTodoLoading(true);
+
+        if (user) {
+          setTempTodo({
+            id: 0,
+            userId: user.id,
+            title,
+            completed: false,
+          });
+
+          const newTodo = await createTodo({
+            userId: user.id,
+            title,
+            completed: false,
+          });
+
+          setTodos(prevTodos => [...prevTodos, newTodo]);
+          setTempTodo(null);
+          setIsNewTodoLoading(false);
+        }
+      } catch (addTodoError) {
+        setErrorMessage('Unable to add a todo');
+      }
+    }, [todos, user],
+  );
+
+  const toggleTodoStatus = useCallback(async (
+    todoId: number,
+    updatedTodoStatus: boolean,
+  ) => {
+    try {
+      const updateResponse = await updateTodoById(
+        { completed: updatedTodoStatus }, todoId,
+      );
+
+      setTodos(prevTodos => (
+        prevTodos.map(todo => {
+          return todo.id === todoId
+            ? { ...todo, completed: updatedTodoStatus }
+            : todo;
+        })
+      ));
+
+      return updateResponse;
+    } catch (updateError) {
+      return setErrorMessage('Unable to update a todo');
+    }
+  }, [todos]);
+
+  const toggleAllTodos = useCallback(async () => {
+    const isAllTodoCompleted = todos.every(todo => todo.completed);
+
+    const updatingTodosIds = todos
+      .filter(todo => (
+        todo.completed === isAllTodoCompleted
+      ))
+      .map(todo => todo.id);
+
+    setUpdatingTodos(updatingTodosIds);
+
+    try {
+      const updatedResponse = await Promise.all(todos.map(todo => (
```

Comment:
You do a lot of work to find updating todos, but still, you run the update on **all** todos. It's not optimal, let's update only needed todos

Reply:
hope I could fix this